### PR TITLE
fix(cast): Use cast platform APIs in MediaCapabilties polyfill

### PR DESCRIPTION
**Pull request by JulianDomingo**
See https://github.com/shaka-project/shaka-player/issues/4726 for more context.

This allows Cast devices to properly filter stream variants with a resolution surpassing that of the device's capabilities.

We place the fix in the `MediaCapabilities` polyfill since it's intended to be the right way to check for anything related to platform support.

--- 

HDR support checks will require `eotf=smpte2048`, as indicated in https://github.com/shaka-project/shaka-player/issues/2813#issue-684874730. Specifically, a `{hev|hvc}1.2` profile is only an *indication* of an HDR transfer function, but *may* be a non-HDR 10-bit color stream. Take for instance this publicly available hls.js HEVC stream (https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_adv_example_hevc/master.m3u8) which contains stream variants with the `hvc1.2.4.L123.B0` codecs, but with a resolution < 4K:

```
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=1996813,BANDWIDTH=2194827,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=960x540,FRAME-RATE=60.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v14/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=6135662,BANDWIDTH=6694727,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=1920x1080,FRAME-RATE=60.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v18/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=4831572,BANDWIDTH=5458398,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=1920x1080,FRAME-RATE=60.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v17/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=3471811,BANDWIDTH=4110269,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=1920x1080,FRAME-RATE=60.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v16/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=2666432,BANDWIDTH=2795200,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=1280x720,FRAME-RATE=60.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v15/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=1169111,BANDWIDTH=1256754,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=768x432,FRAME-RATE=30.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v13/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=859838,BANDWIDTH=932269,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=640x360,FRAME-RATE=30.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v12/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=552728,BANDWIDTH=579426,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=480x270,FRAME-RATE=30.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v11/prog_index.m3u8
#EXT-X-STREAM-INF:AVERAGE-BANDWIDTH=345440,BANDWIDTH=371212,CODECS="hvc1.2.4.L123.B0,ec-3",RESOLUTION=416x234,FRAME-RATE=30.000,CLOSED-CAPTIONS="cc",AUDIO="a3",SUBTITLES="sub1"
v10/prog_index.m3u8
```

In Cast, the platform can distinguish between the two by explicitly providing the transfer function; it uses `smpte2048` (`"PQ"`) because this is the "basis of HDR video formats..." (https://en.wikipedia.org/wiki/Perceptual_quantizer). Below you can see the platform correctly distinguishing between an HDR-capable and non-HDR capable HDMI sink device (using a 4K-capable Chromecast):

`canDisplayType()` results on a 1920x1080 ASUS monitor:
```
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=1920; height=1080; eotf="smpte2084"') => false
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=3840; height=2160; eotf="smpte2084"') => false
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=1920; height=1080') => true
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=3840; height=2160') => false
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"') => true
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0; eotf="smpte2084"') => false
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=1920; height=1080; eotf="smpte2084"') => false
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=3840; height=2160; eotf="smpte2084"') => false
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=1920; height=1080') => true
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=3840; height=2160') => false
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"') => true
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0; eotf="smpte2084"') => false
```

`canDisplayType()` results on a 4K HDR TCL 2017 Roku TV (43"):
```
New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=1920; height=1080; eotf="smpte2084"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=3840; height=2160; eotf="smpte2084"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=1920; height=1080') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"; width=3840; height=2160') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hev1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hev1.2.4.L153.B0; eotf="smpte2084"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=1920; height=1080; eotf="smpte2084"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=3840; height=2160; eotf="smpte2084"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=1920; height=1080') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"; width=3840; height=2160') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0"') => true
VM10:1 New style VP9 codec string must start with vp09. codecId:hvc1.2.4.L153.B0
[JULIAN]: canDisplayType(video/mp4; codecs="hvc1.2.4.L153.B0; eotf="smpte2084"') => true
```

### CODE REVIEW

File: lib/polyfill/media_capabilities.js
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:263]: This is meant to be called only if the platform is cast, so I think instead of returning true here, it should throw an error.

If you remove this check, it would throw on non-cast devices naturally.

I would also suggest you rename this method to include the word "cast" so that the precondition of running on a cast platform were clearer.

File: lib/polyfill/media_capabilities.js
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:239]: "the Cast platform" specifically

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: lib/polyfill/media_capabilities.js
Comment (JulianDomingo) [reply lib/polyfill/media_capabilities.js]: Makes sense, I've added a new error and renamed the function here to mention `Cast`. I also made the cast namespace check `cast`, `cast.__platform__`, and `cast.__platform__.canDisplayType`.

File: lib/polyfill/media_capabilities.js
Comment (JulianDomingo) [reply lib/polyfill/media_capabilities.js]: Done.

**Comment by avelad** [comment]
@JulianDomingo There seems to be a lot of bugs in the tests when running on Chromecast, can you review it? Thanks!

**Comment by JulianDomingo** [comment]
Sure thing, it looks like `canDisplayType()` is returning false for basic video stream configs, surprisingly. So there's no stream variants to test on - looking into it.

File: test/polyfill/media_capabilities_unit.js
Comment (joeyparrish) [inline test/polyfill/media_capabilities_unit.js:190]: Do spyOn calls automatically reset the method to the original after a test case?  I've always assumed they didn't, but this would be really useful if it works that way.

**Review by joeyparrish — APPROVED** [review]
LGTM if spyOn doesn't need cleanup

File: test/polyfill/media_capabilities_unit.js
Comment (JulianDomingo) [reply test/polyfill/media_capabilities_unit.js:190]: It looks like it, based on https://github.com/jasmine/jasmine/issues/683#issuecomment-261629706.

**Comment by JulianDomingo** [comment]
@avelad @joeyparrish Based on my findings with the Chromecast lab failures, the tests were failing due to `cast.__platform__ && cast.__platform__.canDisplayType` being unavailable (but `cast` is available). My solution is to fall back to the original `MediaSource.isTypeSupported(contentType)` if the full cast API set isn't available.

- In this commit which removes the `cast.__platform__ && cast.__platform__.canDisplayType` check (https://github.com/shaka-project/shaka-player/pull/4727/commits/720fe110167f8fece93481f329c555275428dcfa), you can see Chromecast lab tests [fail](https://github.com/shaka-project/shaka-player/actions/runs/3536073542/jobs/5934854261).
- The subsequent commit re-instates the above check (https://github.com/shaka-project/shaka-player/pull/4727/commits/8465a9931374036bc3c9d700c2a3b43dbec09d62) and the Chromecast lab tests [succeed](https://github.com/shaka-project/shaka-player/actions/runs/3536389859).

This obviously contradicts with my manual tests on my Chromecast devices. Maybe this has something to do with the `canDisplayType` binding not being "applied" yet when the tests are ran? I'm not as knowledgeable on how the unit tests / integration tests are "hooked" to the platform when lab tests are ran.

**Review by avelad — APPROVED** [review]

File: lib/polyfill/media_capabilities.js
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:264]: nit: I don't believe you need both of these checks.  I would just say `if (!window.cast)`.

File: lib/polyfill/media_capabilities.js
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:253]: Since all of these come from videoConfig, I feel like you should just pass that instead of creating a new record type.

File: lib/polyfill/media_capabilities.js
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:271]: This is very strange.  I'd like to help you look into that, but I don't want to hold up the PR for it.

File: test/polyfill/media_capabilities_unit.js
Comment (joeyparrish) [reply test/polyfill/media_capabilities_unit.js:196]: That's excellent.  I need to use spyOn more instead of jasmine.createSpy.  We spend too much effort restoring originals.

File: test/polyfill/media_capabilities_unit.js
Comment (joeyparrish) [inline test/polyfill/media_capabilities_unit.js:260]: This looks brittle.  I don't know if we can rely on this exact amount of detail in converting a number to a string on all devices.  Can you make the framerate 24 for this test instead?

**Review by theodab — CHANGES_REQUESTED** [review]

File: lib/polyfill/media_capabilities.js
Comment (JulianDomingo) [reply lib/polyfill/media_capabilities.js:271]: Yeah, was held up on this for a while - the lab tests were failing without this check.

File: lib/polyfill/media_capabilities.js
Comment (JulianDomingo) [reply lib/polyfill/media_capabilities.js:253]: Done, thanks

File: test/polyfill/media_capabilities_unit.js
Comment (JulianDomingo) [reply test/polyfill/media_capabilities_unit.js:260]: Done, forgot to consider precision here.

File: lib/polyfill/media_capabilities.js
Comment (JulianDomingo) [reply lib/polyfill/media_capabilities.js]: Done, removed.

**Review by theodab — APPROVED** [review]

File: lib/polyfill/media_capabilities.js
Comment (joeyparrish) [reply lib/polyfill/media_capabilities.js:248]: I spent a solid day on this, and it's time for me to give up for now.  I managed to forward the Cast platform APIs through an iframe boundary, but not in a way that is compatible with Karma's websocket usage.  :-(

**Review by joeyparrish — APPROVED** [review]

**Pull request merged**
